### PR TITLE
Update combinatorics.jl

### PR DESCRIPTION
**Pull request by matrixmorpheus**
I saw a recent issue about making factorial(n, k) not valid for n<0 or k<0.

This change should allow this.

### CODE REVIEW

**Comment by tkelman** [comment]
Hi, thanks for the PR. It's useful to add a reference to said issue in either the commit message or the description here on GitHub, like so: #9612

**Comment by hayd** [comment]
please add a test for this behaviour!

**Comment by jlapeyre** [comment]
Of course, at a minimum, factorial(-10) and factorial(-10,-10) should be consistent.
i.e. both return a  number, or both throw domain error.

[scipy](http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.misc.factorial.html) returns zero

[mathematica](http://reference.wolfram.com/language/ref/Factorial.html)  never throws domain errors.  if the input is not right the unevaluated form is returned and sometimes (a bit inconsistently) it prints a warning. In the case of factorial, it  automatically uses complex gamma function. This is probably a good choice for Mma.

But, I think the current Julia choice of handling sqrt(-1) is the correct one and should be applied here too. (my 2 cents!)

I don't like this scipy choice, unless there is some good reason. I can't think of mathemtical reason for returning zero. If you interpret n! as complex gamma(n+1), it's still wrong. What use is the zero ? probably nothing, and would only be returned because the input is erroneous. Better to get the error closer to it's origin.  Of course there could be a reason I'm missing, but the scipy page does not mention it.

--John

**Comment by hayd** [comment]
The docs say:
```
Base.factorial(n, k)

   Compute "factorial(n)/factorial(k)"
```
Which reads like this is a shorthand (a more efficient way to calculate) for therefore this must DomainError...

However this leads to the question: what about the result of:

```jl
julia> factorial(3, 4)
0

julia> factorial(3) / factorial(4)
0.25
```
I think this should also error... (not sure what should error type should be thrown though, perhaps an ArgumentError?).

**Comment by jiahao** [comment]
I think it is sensible to ensure that `factorial(n, k) == factorial(n) / factorial(k)` be true even for `n < k`.

**Comment by stevengj** [comment]
The whole point of `factorial(n, k)`, in my understanding, is to compute n!/k! for k ≤ n efficiently and avoiding overflow if possible.  Therefore it should return an integer (for integer inputs), and hence it must throw a `DomainError` for k > n.

**Comment by jiahao** [comment]
@stevengj for comparison, Mathematica's [`FactorialPower` function](http://reference.wolfram.com/language/ref/FactorialPower.html) does almost the same thing  (`factorial(n, k)` in Mathematica is `FactorialPower[n, n-k+1]`) and computes the [descending factorial version of the Pochhammer symbol](https://en.wikipedia.org/wiki/Pochhammer_symbol), which can be continued to almost all complex numbers using the ratio of gamma functions. Ensuring that `factorial(n, k)` is defined for k>n is at least consistent with the latter definition, although arguably the latter definition could be reserved for noninteger k and n arguments.

Another reason in favor of defining `factorial(n, k)` for `n<k` would be to have the identity `factorial(n, k) * factorial(k, n) == 1`.

**Comment by jlapeyre** [comment]
Clearly it exists to compute n!/k! avoiding overflow. But I don't see that this implies that it must return an integer.  There is a legitimate use for  factorial(n,k) with k>n. Suppose I have big n and big k>n. If factorial(n,k) throws a domain error, then I have to write my own.

I suppose its a matter of how much of a PITA it is.

In C I have to write   4 / (double) 5, in Julia I don't.
In Julia I have to write  sqrt(-5 + 0im), in Mathematica, I don't.

It's not 100% clear which is the correct model. I guess writing
1/factorial(k,n) is not too hard if I know k is always greater than n.

**Comment by matrixmorpheus** [comment]
I have added some tests and changed the documentation.

File: test/combinatorics.jl [hunk 1]
```diff
@@ -55,6 +55,9 @@ end
 
 @test factorial(7) == 5040
 @test factorial(7,3) == 7*6*5*4
+@test factorial(3,7) == 0
+@test factorial(-3,-7) == 0
+@test factorial(-7,-3) == 0
```
Comment (hayd) [inline test/combinatorics.jl]: Shouldn't these be

    @test_throws DomainError ...

File: base/combinatorics.jl [hunk 2]
```diff
@@ -62,7 +62,7 @@ end
 # computes n!/k!
 function factorial{T<:Integer}(n::T, k::T)
     if k < 0 || n < 0 || k > n
-        return zero(T)
+        throw(DomainError)
```
Comment (hayd) [inline base/combinatorics.jl]: should be `throw(DomainError())`

File: doc/helpdb.jl [hunk 3]
```diff
@@ -4836,7 +4836,7 @@ popdisplay(d::Display)
 
 ("Base","factorial","factorial(n, k)
 
-   Compute \"factorial(n)/factorial(k)\"
+   Compute \"factorial(n) ÷ factorial(k)\"
```
Comment (hayd) [inline doc/helpdb.jl]: This should probably include doc of behaviour for when it DomainErrors.

**Comment by jiahao** [comment]
Looks like this is superseded by #9824

**Pull request closed without merging**